src/charm.py

```diff
@@ -98,6 +114,75 @@ def _on_loki_pebble_ready(self, _):
     def _on_alertmanager_change(self, _):
         self._configure()
 
+    def _on_ingress_changed(self, _):
+        self._configure()
+        self.loki_provider.update_endpoint(url=self._external_url)
+
+    def _on_logging_relation_changed(self, event):
+        # If there is a change in logging relation, let's update Loki endpoint
+        self.loki_provider.update_endpoint(url=self._external_url, relation=event.relation)
+
+    ##############################################
+    #                 PROPERTIES                 #
+    ##############################################
+
+    @property
+    def _loki_command(self):
+        """Construct command to launch Loki.
+
+        Returns:
+            a string consisting of Loki command and associated
+            command line options.
+        """
+        return f"/usr/bin/loki -config.file={LOKI_CONFIG}"
+
+    @property
+    def _build_pebble_layer(self) -> Layer:
+        """Construct the pebble layer.
+
+        Returns:
+            a Pebble layer specification for the Loki workload container.
+        """
+        pebble_layer = Layer(
+            {
+                "summary": "Loki layer",
+                "description": "pebble config layer for Loki",
+                "services": {
+                    "loki": {
+                        "override": "replace",
+                        "summary": "loki",
+                        "command": self._loki_command,
+                        "startup": "enabled",
+                    },
+                },
+            }
+        )
+
+        return pebble_layer
+
+    @property
+    def hostname(self) -> str:
+        """Unit's hostname."""
+        return socket.getfqdn()
+
+    @property
+    def _external_url(self) -> str:
+        """Return the external hostname to be passed to ingress via the relation."""
+        if ingress_url := self.ingress_per_unit.url:
+            logger.debug("This unit's ingress URL: %s", ingress_url)
+            return ingress_url
+
+        # If we do not have an ingress, then use the pod ip as hostname.
+        # The reason to prefer this over the pod name (which is the actual
+        # hostname visible from the pod) or a K8s service, is that those
+        # are routable virtually exclusively inside the cluster (as they rely)
+        # on the cluster's DNS service, while the ip address is _sometimes_
+        # routable from the outside, e.g., when deploying on MicroK8s on Linux.
+        return f"http://{self.hostname}:{self._port}"
```

Comment:
The comment seems to say we us IP address but the value set is hostname. Which one is it supposed to be? I would recommend hostname.

Reply:
You are right, hostname.